File: AndroidApp/src/main/java/com/smp/sharedkotlin/app/data/MemoDataSourceImpl.kt
package com.smp.sharedkotlin.app.data

import android.util.Log
import java.util.*

class MemoDataSourceImpl : MemoDataSource {

    companion object {
        const val MOCK_NETWORK_DELAY = 1500L
        const val MAX_MOCK_MEMO = 100
    }

    private val mockMemos: List<MemoModel> by lazy {
        val cal = Calendar.getInstance()

        1.rangeTo(MAX_MOCK_MEMO).map {
            MemoModel(
                it,
                "Test Memo $it",
                "This is test memo $it",
                cal.apply { add(Calendar.DAY_OF_YEAR, -1) }.timeInMillis,
                cal.apply { add(Calendar.DAY_OF_YEAR, -1) }.timeInMillis
            )
        }
    }

    override fun getAllMemos(): List<MemoModel> {
        try {
            Thread.sleep(MOCK_NETWORK_DELAY)
            return mockMemos.toList()
        } catch (throwable: Throwable) {
            Log.e("MemoDataSource", "getMemoList Error : $throwable")
        }
        return listOf()
    }

    override fun getMemo(id: Int): MemoModel? {
        if (id <= 0 || id > MAX_MOCK_MEMO) {
            return null
        }

        try {
            Thread.sleep(MOCK_NETWORK_DELAY)
            return mockMemos[id - 1]
        } catch (throwable: Throwable) {
            Log.e("MemoDataSource", "getMemo Error : $throwable")
        }
        return null
    }

}

File: iOSApp/iOSApp/Data/MemoDataSourceImpl.swift
import Foundation


class MemoDataSourceImpl: MemoDataSource {
    
    static let MOCK_NETWORK_DELAY: UInt32 = 1 //sec
    static let MAX_MOCK_MEMO: Int32 = 100
    
    let calender = Calendar.current
    
    lazy var mockMemos: [MemoModel] = {
        (1..<MemoDataSourceImpl.MAX_MOCK_MEMO).map {
            let date = self.calender.date(byAdding: .day, value: (-1 * Int($0)), to: Date())
            let timeInterval = Int64(date?.timeIntervalSince1970 ?? 0)

            return MemoModel(id: $0, title: "Test Memo \($0)", content: "This is test memo \($0)", createdAt: timeInterval, lastModifiedAt: timeInterval)
        }
    } ()
    
    func getAllMemos() -> [MemoModel] {
        sleep(MemoDataSourceImpl.MOCK_NETWORK_DELAY)
        return self.mockMemos
    }
    
    func getMemo(id: Int32) -> MemoModel? {
        
        if (id <= 0 || id > MemoDataSourceImpl.MAX_MOCK_MEMO) {
            return nil
        }
        
        sleep(MemoDataSourceImpl.MOCK_NETWORK_DELAY)
        return self.mockMemos[Int(id) - 1]
    }
}
